Add Carnot cycle efficiency exercise

tommasodiotalevi opened the pull request:
This pull request adds a new Carnot module to the project.
The new module:
- calculates the efficiency of the Carnot cycle, computing work and heat exchanges for all the transformations;
- compares it with the ideal Carnot efficiency operating between two temperature reservoirs;

Objectives:
- Compute the efficiency of a Carnot engine, for all the thermodynamic transformations (isothermal and adiabatic);
- Show that such efficiency converges to the ideal Carnot formula, when the number of intermediate steps gets larger (quasistatic conditions).

tommasodiotalevi committed:
Add readme

tommasodiotalevi committed:
Add Carnot exercise

tommasodiotalevi committed:
Add project to CMakeLists

tommasodiotalevi committed:
Fix variable redefinition

Error visible only with GCC and not in Apple Clang

giacomini commented on problemi/carnot/carnot.cpp:
```diff
@@ -0,0 +1,168 @@
+
+ public:
+  State(double V, double T)
+      : V_(V), T_(T), P_((R * T) / V) {}  // Costruttore di stato gas "ideale".
```
```suggestion
      : V_(V), T_(T), P_((R * T) / V) {}
```
se il commento e' veramente necessario, e' meglio metterlo non in fondo alla riga, altrimenti la formattazione soffre

giacomini commented on problemi/carnot/carnot.cpp:
```diff
@@ -0,0 +1,168 @@
+  double V() const { return V_; }
+  double T() const { return T_; }
+  double P() const { return P_; }
+  void setP(double P) { P_ = P; }
```
se ci sono anche i setter, e' inutile avere una classe, non c'e' modo di mantenere l'invariante di classe

tommasodiotalevi committed:
Improve readme

tommasodiotalevi committed:
Improve State design and error handling

This commit improves the design of the State class, adding constraints
on the thermodynamic state variables and removing inconsistencies.
Minor error handling also for the integration procedure.
